File: python/tests/test_gosling.py
import pytest

from bioviz import Gosling


def test_gosling_passes_spec_through_verbatim():
    spec = {
        "title": "t",
        "tracks": [{"mark": "bar", "x": {"field": "start", "type": "genomic"}}],
    }
    w = Gosling(spec, padding=30, theme="dark")

    assert w.options["spec"] == spec
    assert w.options["padding"] == 30
    assert w.options["theme"] == "dark"
    # Gosling is spec-driven, so no columns are packed.
    assert w.data["columns"] == {}
    assert len(w.buffer) == 0


def test_gosling_omits_padding_and_theme_when_absent():
    w = Gosling({"tracks": []})
    assert "padding" not in w.options
    assert "theme" not in w.options


def test_gosling_copies_the_spec():
    spec = {"tracks": []}
    w = Gosling(spec)
    spec["tracks"].append("mutated")
    # The widget holds its own copy of the top-level mapping.
    assert w.options["spec"] is not spec


def test_gosling_validates_input():
    with pytest.raises(ValueError, match="must be a mapping"):
        Gosling("not a spec")
    with pytest.raises(ValueError, match="at least one of"):
        Gosling({"foo": 1})

File: pkg-r/tests/testthat/test-gosling.R
test_that("gosling() builds an htmlwidget carrying the spec verbatim", {
  spec <- list(
    title = "t",
    tracks = list(list(
      mark = "bar",
      x = list(field = "start", type = "genomic")
    ))
  )
  w <- gosling(spec, padding = 30, theme = "dark")

  expect_s3_class(w, "htmlwidget")
  # The spec is passed through under options$spec, unmodified.
  expect_equal(w$x$options$spec, spec)
  expect_equal(w$x$options$padding, 30)
  expect_equal(w$x$options$theme, "dark")
  # Gosling is spec-driven, so columns are empty.
  expect_equal(w$x$data$columns, list())
})

test_that("gosling() omits padding and theme when not supplied", {
  w <- gosling(list(tracks = list()))
  expect_null(w$x$options$padding)
  expect_null(w$x$options$theme)
})

test_that("gosling() validates its input", {
  expect_error(gosling("not a list"), "must be a list")
  expect_error(gosling(list(foo = 1)), "must contain at least one")
})
